### PR TITLE
Remove unnecessary updates, recommended packages

### PR DESCRIPTION
**Pull request by alayek**
Things changed in this PR
- Expand on heroku install, to remove unnecessary `apt-get update`.
- Add `--no-install-recommends` flag to not installed recommended packages
- Add `man` install
- Add test instructions on README

I have tested on my Linux and Windows.

### CODE REVIEW

**Comment by byteknacker** [comment]
Testing now on Mac OS 10.11.4

File: provision.sh [hunk 1]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
 
+# install heroku toolbelt
+echo "-------------- Installing heroku toolbelt -------------------------"
+#wget -O- https://toolbelt.heroku.com/install-ubuntu.sh | sh
+# These shell script snippets are directly taken from heroku installation script
+# We want to avoid the apt-get update
+# add heroku repository to apt
+echo "deb http://toolbelt.heroku.com/ubuntu ./" > /etc/apt/sources.list.d/heroku.list
+# install heroku's release key for package verification
+wget -O- https://toolbelt.heroku.com/apt/release.key | apt-key add -
```
Comment (byteknacker) [inline provision.sh:25]: What is the apt-key? I don't understand this option. Is it an option or another piped command?

File: provision.sh [hunk 2]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
 echo "Running vagrant provisioning"
 echo "---------------------------------------------"
 
+# install heroku toolbelt
+echo "-------------- Installing heroku toolbelt -------------------------"
+#wget -O- https://toolbelt.heroku.com/install-ubuntu.sh | sh
+# These shell script snippets are directly taken from heroku installation script
+# We want to avoid the apt-get update
+# add heroku repository to apt
+echo "deb http://toolbelt.heroku.com/ubuntu ./" > /etc/apt/sources.list.d/heroku.list
```
Comment (byteknacker) [inline provision.sh:23]: Why don't you wish to have the apt-get update? Isn't it good practice? Because it will install the latest versions of everything else below this update.

File: provision.sh [hunk 3]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
+# install heroku's release key for package verification
+wget -O- https://toolbelt.heroku.com/apt/release.key | apt-key add -
+
 echo "------------- Updating packages and their dependencies -----------"
 apt-get update -y # no need for sudo, and -y is needed to bypass the yes-no
```
Comment (byteknacker) [inline provision.sh:28]: I thought you deleted this line

File: provision.sh [hunk 4]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
+
+# install the toolbelt
+apt-get install heroku-toolbelt --no-install-recommends -y
```
Comment (byteknacker) [inline provision.sh:31]: @alayek how many MB did you save with this. What do the recommended packages include? Will it negatively influence a typical project?

File: provision.sh [hunk 5]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
+
 echo "-------------- Installing build-essential and pip ----------------"
-apt-get install -y build-essential python-pip
```
Comment (byteknacker) [inline provision.sh:20]: Why did you delete them? Are they already included in the base box?

File: provision.sh [hunk 6]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
+# install gcc and g++ and other build basics to ensure most software works
+# install man too
+apt-get install build-essential python-pip man -y --no-install-recommends
```
Comment (byteknacker) [inline provision.sh:36]: Sorry got it @alayek, did not see this line before.

File: provision.sh [hunk 7]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
 
-# install heroku toolbelt
-echo "-------------- Installing heroku toolbelt -------------------------"
-wget -O- https://toolbelt.heroku.com/install-ubuntu.sh | sh
```
Comment (byteknacker) [inline provision.sh:24]: What is that | sh?

File: provision.sh [hunk 8]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
+# needed for heroku toolbelt
+# notice that this is not a rigorous Ruby install, where we typically use rvm
+apt-get install ruby --no-install-recommends -y
```
Comment (byteknacker) [inline provision.sh:40]: Is this really needed? Cause I have not run into troubles without this. I can use Heroku normally without installing ruby. I thought it was already preinstalled on the base box.

File: provision.sh (hunk 1 above)
Comment (alayek) [inline provision.sh:25]: Earlier we were doing installation via heroku. We hit a URL, it gives out a shell script, and we execute it.

I looked into the script, deleted unnecessary stuff, kept the stuff that's actually needed.

File: provision.sh [hunk 9]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
 
 # install virtualenv and virtualenvwrapper
 echo "--------------- Installing both virtualenv and virtualenvwrapper ---"
 pip install virtualenvwrapper virtualenv
 
 # install postgresql and setup user
 echo "--------------- Installing postgresql ------------------------------"
-apt-get install -y python-dev python3-dev libpq-dev postgresql postgresql-contrib
+apt-get install python-dev python3-dev libpq-dev postgresql postgresql-contrib --no-install-recommends -y
```
Comment (byteknacker) [inline provision.sh:48]: Do you put -y at the very end or do you put it after install? When I read it on SO, I thought the latter version is correct.

Comment (byteknacker) [inline provision.sh:48]: Also do we really need python-dev if we already have python3-dev?

File: provision.sh [hunk 10]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
 su - postgres -c "createuser -s vagrant"
 su - vagrant -c "createdb ${DB}"
 
 echo "--------------- Preparing .bashrc for first usage ------------------"
 # set up virtualenvwrapper
 echo WORKON_HOME="/home/vagrant/.virtualenvs" >> /home/vagrant/.bashrc
 # activate virtualenvwrapper
-echo "source /usr/local/bin/virtualenvwrapper.sh" >> /home/vagrant/.bashrc
 echo "WORKON_HOME=\"/home/vagrant/.virtualenvs\"" >> /home/vagrant/.bashrc
```
Comment (byteknacker) [inline provision.sh:56]: Why is this line not redundant to the line 54? There is just an extra quote between the string of the command echo.

File: provision.sh [hunk 11]
```diff
@@ -14,31 +14,47 @@ echo "---------------------------------------------"
...
+echo "source /usr/local/bin/virtualenvwrapper.sh > /dev/null 2>&1" >> /home/vagrant/.bashrc
```
Comment (byteknacker) [inline provision.sh:57]: What does this line do? 57

File: provision.sh (hunk 1 above)
Comment (alayek) [inline provision.sh:25]: The key addition was part of earlier install. It's standard practice to do a key-based verification that what you are installing is indeed what it claims to be.

Check `rvm` (Ruby version manager) install for example. It starts with adding a key to your machine so that you can verify it.

Similar stuff is also done under the hood by package managers or build tools like Maven for Java, or pip for Python.

File: provision.sh (hunk 2 above)
Comment (alayek) [inline provision.sh:23]: `apt-get update` does not install anything. That's `apt-get upgrade`.

In general terms, `apt-get update` updates the list of installed packages and dependencies. It hits a lot of package servers and finds out latest information about what new package versions have been released for all installed software.

Then, it goes on and updates the list of dependencies properly on your machine.

Here, `apt-get update` is needed, to update its internal lists and storages to inform of heroku-toolbelt. But we run it once.

Because it's a time consuming operation. In its first run itself, it would pick up things necessary for next installations.

File: provision.sh (hunk 4 above)
Comment (alayek) [inline provision.sh:31]: It doesn't save in terms of MB. I didn't check, there might be some 1 MB savings or so. But it saves in reduced number of requests to package servers. It saves somewhat in time.

File: provision.sh (hunk 1 above)
Comment (byteknacker) [inline provision.sh:25]: Thanks for the information @alayek. Where can you learn this stuff more properly? Do you have any recommendations? Is FCC good for shell scripting and system admin?

File: provision.sh (hunk 2 above)
Comment (byteknacker) [inline provision.sh:23]: So it is best practice to put apt-get update at the middle/end of a shell script if there are follow up installations? Why didn't you put it at the very end?

File: provision.sh (hunk 1 above)
Comment (alayek) [inline provision.sh:25]: Umm...I didn't learn this from FCC though. I learned by working with systems, failing and learning by Googling. No worries. If you keep asking questions, you would learn a lot :+1:

File: provision.sh (hunk 2 above)
Comment (alayek) [inline provision.sh:23]: I usually run `apt-get update` after updating the `sources.list` and before installing something else. This something else might or might not depend on my recent change to the `sources.list`.

I don't know if that doe or does not, and that decision is better left to package manager. So, I run it them to be sure what I install next won't cause any problem with what I already have.

**Comment by byteknacker** [comment]
@alayek I am good to go. Tested everything I can think of, unless you wish me to test something more, I shall merge this.

**Comment by alayek** [comment]
@koustuvsinha only need your approval :grin:

**Comment by koustuvsinha** [comment]
# LGTM 👍
